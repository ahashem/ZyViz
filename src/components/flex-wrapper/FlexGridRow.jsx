import React from 'react';
import PropTypes from 'prop-types';
import uuid from 'uuid/v4';
import { Col, Row } from 'antd';

const AntdBoxes = 24; // By Design in AntD (24 boxes - 12 Columns)

/**
 * Generate `Col` wrapper with calculated size according to number of wrapped elements (i.e: `props.children`)
 * @param children
 * @return {*[]}
 * @constructor
 */
const ChildrenGen = (children) => {
  const childrenMap = React.Children.toArray(children);
  const childrenCount = React.Children.count(children);

  const boxByCountSize = ((AntdBoxes / childrenCount) >= AntdBoxes) ? AntdBoxes : AntdBoxes / childrenCount;

  return childrenMap.map((child, index) => {
    return (
      <Col
        key={uuid()}
        xs={AntdBoxes} sm={AntdBoxes}
        md={boxByCountSize}
        span={boxByCountSize}
        lg={boxByCountSize}
      >
        {child}
      </Col>
    );
  });
};

const FlexGridRow = ({ gutter, justify, children }) => {
  return (
    <Row
      type="flex"
      justify={justify}
      gutter={gutter || AntdBoxes}
    >
      {ChildrenGen(children)}
    </Row>
  );
};

FlexGridRow.propTypes = {
  gutter: PropTypes.number,
  justify: PropTypes.string,
  children: PropTypes.node,
};

FlexGridRow.defaultProps = {
  gutter: AntdBoxes,
  justify: 'space-around',
  children: null,
};

export default FlexGridRow;
