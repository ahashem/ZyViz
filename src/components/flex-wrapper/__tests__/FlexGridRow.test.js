import React from 'react';
import { shallow } from 'enzyme';
import FlexGridRow from '../FlexGridRow';

describe('<FlexGridRow /> component', () => {
  it('should render `Col` according to children nodes count and single `Row`', () => {
    const children = Array(4).fill(<div/>);

    const rendered = shallow(
      <FlexGridRow>
        {children}
      </FlexGridRow>
    );

    expect(rendered.find('Row')).toHaveLength(1);
    expect(rendered.find('Col')).toHaveLength(4);
  });

});

