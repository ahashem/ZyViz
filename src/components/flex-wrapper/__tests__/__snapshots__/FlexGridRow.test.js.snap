// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FlexGridRow /> component should render \`Col\` according to children nodes count and single \`Row\` 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <FlexGridRow
    gutter={24}
    justify="space-around"
  >
    <div />
    <div />
    <div />
    <div />
  </FlexGridRow>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Col
          lg={6}
          md={6}
          sm={24}
          span={6}
          xs={24}
        >
          <div />
        </Col>,
        <Col
          lg={6}
          md={6}
          sm={24}
          span={6}
          xs={24}
        >
          <div />
        </Col>,
        <Col
          lg={6}
          md={6}
          sm={24}
          span={6}
          xs={24}
        >
          <div />
        </Col>,
        <Col
          lg={6}
          md={6}
          sm={24}
          span={6}
          xs={24}
        >
          <div />
        </Col>,
      ],
      "gutter": 24,
      "justify": "space-around",
      "type": "flex",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "b0e1cb93-2ef2-42c8-a218-70cc25898936",
        "nodeType": "class",
        "props": Object {
          "children": <div />,
          "lg": 6,
          "md": 6,
          "sm": 24,
          "span": 6,
          "xs": 24,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": ".0",
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "ecd80305-5f95-4d34-8bf5-a5d9f0031382",
        "nodeType": "class",
        "props": Object {
          "children": <div />,
          "lg": 6,
          "md": 6,
          "sm": 24,
          "span": 6,
          "xs": 24,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": ".1",
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "1fba8ea4-f594-4098-85c1-cc73241f65f2",
        "nodeType": "class",
        "props": Object {
          "children": <div />,
          "lg": 6,
          "md": 6,
          "sm": 24,
          "span": 6,
          "xs": 24,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": ".2",
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "da71cffe-0365-47f2-9cac-a1974a5a9f8c",
        "nodeType": "class",
        "props": Object {
          "children": <div />,
          "lg": 6,
          "md": 6,
          "sm": 24,
          "span": 6,
          "xs": 24,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": ".3",
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "div",
        },
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Col
            lg={6}
            md={6}
            sm={24}
            span={6}
            xs={24}
          >
            <div />
          </Col>,
          <Col
            lg={6}
            md={6}
            sm={24}
            span={6}
            xs={24}
          >
            <div />
          </Col>,
          <Col
            lg={6}
            md={6}
            sm={24}
            span={6}
            xs={24}
          >
            <div />
          </Col>,
          <Col
            lg={6}
            md={6}
            sm={24}
            span={6}
            xs={24}
          >
            <div />
          </Col>,
        ],
        "gutter": 24,
        "justify": "space-around",
        "type": "flex",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "b0e1cb93-2ef2-42c8-a218-70cc25898936",
          "nodeType": "class",
          "props": Object {
            "children": <div />,
            "lg": 6,
            "md": 6,
            "sm": 24,
            "span": 6,
            "xs": 24,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": ".0",
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "ecd80305-5f95-4d34-8bf5-a5d9f0031382",
          "nodeType": "class",
          "props": Object {
            "children": <div />,
            "lg": 6,
            "md": 6,
            "sm": 24,
            "span": 6,
            "xs": 24,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": ".1",
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "1fba8ea4-f594-4098-85c1-cc73241f65f2",
          "nodeType": "class",
          "props": Object {
            "children": <div />,
            "lg": 6,
            "md": 6,
            "sm": 24,
            "span": 6,
            "xs": 24,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": ".2",
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "da71cffe-0365-47f2-9cac-a1974a5a9f8c",
          "nodeType": "class",
          "props": Object {
            "children": <div />,
            "lg": 6,
            "md": 6,
            "sm": 24,
            "span": 6,
            "xs": 24,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": ".3",
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
